### DISCUSSION TITLE
Deterministic zipped string - how?

### DISCUSSION THREAD

**Discussion by oyevteyev** [icsharpcode/SharpZipLib#752] [Q&A]
We do compress long strings into a zipped one (password protected).
Our problem is - the output is not deterministic - every time we compress a string - we are getting a different compressed sting.
Unzipping it always returns the same result .
How to make it deterministic - the same zipped string every time?
That will help us with caching.
Thanks.
Zip:

        public string Zip(string clearText, string password)
        {
            if (clearText == null)
            {
                return null;
            }          
            using var memOutput = new MemoryStream();
            using var zipOut = new ZipOutputStream(memOutput);
            zipOut.Password = password;
            zipOut.SetLevel(3);
            var entry = new ZipEntry("state");
            zipOut.PutNextEntry(entry);
            using var writer = new StreamWriter(zipOut);
            writer.Write(clearText);
            writer.Flush();
            zipOut.Finish();
            var bytes = new byte[memOutput.Length];
            memOutput.Seek(0, SeekOrigin.Begin);
            var _ = memOutput.Read(bytes, 0, bytes.Length);
            return Convert.ToBase64String(bytes);
        }

Unzip:

      public string UnZip(string text, string password, bool isBase64Encoded = true)
        {
            if (text == null)
            {
                return null;
            }
            try
            {
                var bytes = isBase64Encoded ? Convert.FromBase64String(text) : Encoding.Unicode.GetBytes(text);
                using var memInput = new MemoryStream(bytes);
                using var zipInput = new ZipInputStream(memInput);
                zipInput.Password = password;
                zipInput.GetNextEntry();
                using var reader = new StreamReader(zipInput);
                var txt = reader.ReadToEnd();
                return txt;
            }
            catch
            {
                return "[]";
            }
        }

**Comment by piksel** [comment]
The modified time in the file is always different. Try setting the modified time manually for the entry.

**Reply by oyevteyev** [reply]
Tried, but the entry is still different each time

        public string Zip(string clearText, string password)
        {
            if (clearText == null)
            {
                return null;
            }

            var dt = new DateTime(2022, 1, 1);
            using var memOutput = new MemoryStream();
            using var zipOut = new ZipOutputStream(memOutput);
            zipOut.Password = password;
            zipOut.SetLevel(3);
            var entry = new ZipEntry("state")
            {
                DateTime = dt,
                DosTime = dt.Ticks
            };
            zipOut.PutNextEntry(entry);
            using var writer = new StreamWriter(zipOut);
            writer.Write(clearText);

            writer.Flush();
            zipOut.Finish();

            var bytes = new byte[memOutput.Length];
            memOutput.Seek(0, SeekOrigin.Begin);
            var _ = memOutput.Read(bytes, 0, bytes.Length);

            return Convert.ToBase64String(bytes);
        }

**Comment by AqlaSolutions** [comment]
That's because encryption is not deterministic, e.g. AES uses random IV each time you run it. I solved it here: https://github.com/AqlaSolutions/DropboxEncryptedUploader/blob/3e6530b3944c41156884583fcd819848e1ea6512/Infrastructure/ZipEncryptionHelper.cs
